k2/python/k2/mwer_loss.py

```diff
@@ -0,0 +1,177 @@
+            'sum': sum loss of each path over the whole batch together.
+            'mean': divide above 'sum' by total num paths over the whole batch.
+        '''
+
+        assert reduction in ('none', 'mean', 'sum'), reduction
+        super().__init__()
+        self.temperature = temperature
+        self.use_double_scores = use_double_scores
+        self.reduction = reduction
+
+    def forward(self,
+                lattice: k2.Fsa,
+                ref_texts: Union[k2.RaggedTensor, List[List[int]]],
+                nbest_scale: float,
+                num_paths: int) -> Union[torch.Tensor, k2.RaggedTensor]:
+        '''Compute the Minimum Word Error loss given
+        a lattice and corresponding ref_texts.
+
+        Args:
+          lattice:
+            An FsaVec with axes [utt][state][arc].
+          ref_texts:
+            It can be one of the following types:
+              - A list of list-of-integers, e..g, `[ [1, 2], [1, 2, 3] ]`
+              - An instance of :class:`k2.RaggedTensor`.
+                Must have `num_axes == 2` and with dtype `torch.int32`.
+          nbest_scale:
+            Scale `lattice.score` before passing it to :func:`k2.random_paths`.
+            A smaller value leads to more unique paths at the risk of being not
+            to sample the path with the best score.
+          num_paths:
+            Number of paths to **sample** from the lattice
+            using :func:`k2.random_paths`.
+        Returns:
+            Minimum Word Error Rate loss.
+        '''
+
+        nbest = k2.Nbest.from_lattice(
+            lattice=lattice,
+            num_paths=num_paths,
+            use_double_scores=self.use_double_scores,
+            nbest_scale=nbest_scale,
+        )
+        device = lattice.scores.device
+        path_arc_shape = nbest.kept_path.shape.to(device)
+        stream_path_shape = nbest.shape.to(device)
+
+        hyps = nbest.build_levenshtein_graphs()
+        refs = k2.levenshtein_graph(ref_texts, device=hyps.device)
+        levenshtein_alignment = k2.levenshtein_alignment(
+            refs=refs,
+            hyps=hyps,
+            hyp_to_ref_map=nbest.shape.row_ids(1),
+            sorted_match_ref=True,
+        )
+        tot_scores = levenshtein_alignment.get_tot_scores(
+            use_double_scores=self.use_double_scores, log_semiring=False
+        )
+        # Each path has a corresponding wer.
+        wers = -tot_scores.to(device)
+
+        # Group each log_prob into [path][arc]
+        ragged_nbest_logp = k2.RaggedTensor(path_arc_shape, nbest.fsa.scores)
+        # Get the probalitity of each path, in log format,
+        # with shape [stream][path].
+        path_logp = ragged_nbest_logp.sum() / self.temperature
+        ragged_path_prob = k2.RaggedTensor(stream_path_shape, path_logp.exp())
+
+        # Normalize prob of each path.
+        den_prob = ragged_path_prob.sum()
+        den_logp = torch.index_select(
```

Comment:
For some versions of PyTorch, e.g., torch 1.7.1, the index type has to be `torch.int64`, I think.